crypto/fipsmodule/evp/p_hmac.c

```diff
@@ -0,0 +1,230 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OpenSSL PROJECT OR
+ * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ * ====================================================================
+ *
+ * This product includes cryptographic software written by Eric Young
+ * (eay@cryptsoft.com).  This product includes software written by Tim
+ * Hudson (tjh@cryptsoft.com). */
+
+#include <openssl/evp.h>
+
+#include <openssl/asn1.h>
+#include <openssl/err.h>
+#include <openssl/hmac.h>
+#include <openssl/mem.h>
+#include <openssl/obj.h>
+
+#include "../../evp_extra/internal.h"
+#include "../digest/internal.h"
+#include "internal.h"
+
+
+typedef struct {
+  const EVP_MD *md; /* MD for HMAC use */
+  uint8_t *key;
+  size_t key_len;
+  HMAC_CTX ctx;
+} HMAC_PKEY_CTX;
+
+static int hmac_init(EVP_PKEY_CTX *ctx) {
+  HMAC_PKEY_CTX *hctx;
+  hctx = OPENSSL_malloc(sizeof(HMAC_PKEY_CTX));
+  if (hctx == NULL) {
+    return 0;
+  }
+  OPENSSL_memset(hctx, 0, sizeof(HMAC_PKEY_CTX));
+  HMAC_CTX_init(&hctx->ctx);
+  ctx->data = hctx;
+  return 1;
+}
+
+static int hmac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src) {
+  HMAC_PKEY_CTX *sctx, *dctx;
+  if (!hmac_init(dst)) {
+    return 0;
+  }
+  sctx = src->data;
+  dctx = dst->data;
+  dctx->md = sctx->md;
+  HMAC_CTX_init(&dctx->ctx);
+  if (!HMAC_CTX_copy_ex(&dctx->ctx, &sctx->ctx)) {
+    return 0;
+  }
+
+  if (sctx->key_len != 0) {
+    dctx->key = OPENSSL_memdup(sctx->key, sctx->key_len);
+    if (dctx->key == NULL) {
+      return 0;
+    }
+    dctx->key_len = sctx->key_len;
+  }
+  return 1;
+}
+
+static void hmac_cleanup(EVP_PKEY_CTX *ctx) {
+  HMAC_PKEY_CTX *hctx = ctx->data;
+  if (hctx == NULL) {
+    return;
+  }
+  HMAC_CTX_cleanse(&hctx->ctx);
+  if (hctx->key != NULL) {
+    OPENSSL_free(hctx->key);
+  }
+  OPENSSL_free(hctx);
+}
+
+static int hmac_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  HMAC_PKEY_CTX *hctx = ctx->data;
+  CBB *key = OPENSSL_malloc(sizeof(CBB));
+  if (key == NULL) {
+    return 0;
+  }
+
+  if (!CBB_init(key, 0) || !CBB_add_bytes(key, hctx->key, hctx->key_len)) {
+    CBB_cleanup(key);
+    return 0;
+  }
+
+  // |EVP_PKEY_new_mac_key| allocates a temporary |EVP_PKEY_CTX| and frees it
+  // along with |ctx->data| (which contains the original key data).
+  // We allocate and copy the key data over to a new |CBB| and assign it to
+  // |pkey|, so that it is available in later operations.
+  return EVP_PKEY_assign(pkey, EVP_PKEY_HMAC, key);
```

Comment:
Is there a better type for the key? In other EVP_PKEYs we use the `RSA`, `KEM_KEY`, or `EC_KEY`. This feels a little weird to pass around a key in a CBB. Does something else in the existing HMAC code expect to use a CBB for the key?